### PR TITLE
Pull request to advance the implementation of the command message

### PR DESCRIPTION
**Pull request by adamrankin**
Correcting CMake error to only build v3 messages when v3 is requested

Adding RTTI option for message type identification

Renaming a couple of functions to better match there behavior. This change is not backward compatible so I am open to discussion. I suppose we could have both function names to not break existing codebases but my opinion is that its better to rip off the bandage once.

### CODE REVIEW

File: Source/igtlMessageFactory.cxx [hunk 1]
```diff
@@ -87,54 +91,60 @@ igtl::MessageBase::Pointer MessageFactory::GetMessage(igtl::MessageHeader::Point
...
     std::ostringstream oss;
-    oss << "Error: In " __FILE__ ", line " << __LINE__ << ", receiving message of type:" << headerMsg->GetDeviceType() << std::endl;
+    oss << "Error: In " __FILE__ ", line " << __LINE__ << ", receiving message of type:" << headerMsg->GetType() << std::endl;
     throw std::invalid_argument(oss.str());
   }
 
   igtl::MessageBase::Pointer result = NULL;
 
-  if (strcmp(headerMsg->GetDeviceType(), "TRANSFORM") == 0)
+  if ( headerMsg->GetType().compare(std::string("TRANSFORM")) == 0)
```
Comment (christiana) [inline Source/igtlMessageFactory.cxx]: If OpenIGTLink_RTTI was always on, GetIGTLMessageType() could be used here, right?

i.e.

  headerMsg->GetType().compare(TransformMessage::GetIGTLMessageType()) == 0)

I cannot see that the RTTI is used anywhere, so I wonder why it is added.

Comment (adamrankin) [inline Source/igtlMessageFactory.cxx]: You're correct, I implemented the RTTI after this change. I will go back and change this!

The RTTI is more for client/server implementations that need to parse message types.

See https://www.assembla.com/spaces/plus/tickets/1057-upgrade-plus-openigtlink-to-new-v3-specification/details for usage

File: Source/igtlBindMessage.cxx [hunk 2]
```diff
@@ -58,7 +58,7 @@ int BindMessageBase::AppendChildMessage(igtl::MessageBase * child)
   if (this->m_ChildMessages.size() < 0xFFFF)
     {
     ChildMessageInfo info;
-    info.type = child->GetDeviceType();
+    info.type = child->GetType();
```
Comment (lassoan) [inline Source/igtlBindMessage.cxx]: Was this change intentional?
DeviceType and DeviceName method names are nice and consistent.

File: CMakeLists.txt [hunk 3]
```diff
@@ -15,6 +15,8 @@ SET(OpenIGTLink_VERSION_PATCH "0")
 option(OpenIGTLink_PROTOCOL_VERSION_2 "Build Library for Protocol Version 2" ON)
 option(OpenIGTLink_PROTOCOL_VERSION_3 "Build Library for Protocol Version 3" OFF)
 
+option(OpenIGTLink_MessageRTTI "Enable RTTI information in message classes" OFF)
```
Comment (lassoan) [inline CMakeLists.txt]: RTTI is a well-defined C++ feature. We should not hijack it and use it for something else (that has goal but a completely different implementation with different capabilities).
What is this RTTI-replacement mechanism needed for?

File: Source/igtlCommandMessage.cxx [hunk 4]
```diff
@@ -42,19 +41,52 @@ namespace igtl {
...
+  if (strlen(aCommandName) > IGTL_COMMAND_NAME_SIZE) /* If the length is beyond the range specified by the spec */
+  {
+    return 0;
+  }
+  strcpy((char*)m_CommandName, aCommandName);
+  return 1;
+}
+
+
+int CommandMessage::SetCommandName(const std::string& aCommandName)
```
Comment (lassoan) [inline Source/igtlCommandMessage.cxx]: this should just call the const char* version (if needed at all)

File: Source/igtlMessageHandler.h [hunk 5]
```diff
@@ -36,7 +36,7 @@ class IGTLCommon_EXPORT MessageHandler: public Object
   igtlNewMacro(igtl::MessageHandler);
 
  public:
-  virtual const char* GetMessageType() { return ""; }
+  virtual std::string GetMessageType() const { return std::string(""); }
```
Comment (lassoan) [inline Source/igtlMessageHandler.h]: This change is probably not backward compatible. You might not get a build error but you'll get memory corruption if you store the result in a const char* as before (the std::string is deleted when execution goes to the next line; therefore the const char* becomes a dangling pointer).

File: Source/igtlutil/igtl_command.h [hunk 6]
```diff
@@ -29,9 +30,11 @@ extern "C" {
 #pragma pack(1)     /* For 1-byte boundary in memory */
 
 typedef struct {
-  igtl_uint16    encoding;         /* Character encoding type as MIBenum value (defined by IANA). Default=3. */
+  igtl_uint32    commandId;        /* The unique ID of this command - 4 bytes*/
+  igtl_uint8     commandName[32];  /* The name of this command - 32 bytes*/
+  igtl_uint16    encoding;         /* Character encoding type as MIBenum value (defined by IANA). Default=3 - 2 bytes*/
                                    /* Please refer http://www.iana.org/assignments/character-sets for detail */
-  igtl_uint32    length;           /* Length of command */
+  igtl_uint32    length;           /* Length of command - 4 bytes */
```
Comment (lassoan) [inline Source/igtlutil/igtl_command.h]: why the -4 bytes?

File: CMakeLists.txt (hunk 3 above)
Comment (christiana) [inline CMakeLists.txt]: I agree. RTTI (https://en.wikipedia.org/wiki/Run-time_type_information) is a C++ feature, so the feature added here needs a different name.

From what I understand, the static function GetIGTLMessageType() returns the same value as GetType() (GetDeviceType()). It can be useful to have this text string available both statically and dynamically, but then GetType() should call GetIGTLMessageType() (or do it indirectly, via the field m_DefaultBodyType).

File: Source/igtlBindMessage.cxx (hunk 2 above)
Comment (christiana) [inline Source/igtlBindMessage.cxx]: The two existing methods GetDeviceName() and GetDeviceType() written to be symmetrical, but refer to quite different values. DeviceName is a user-defined name for the device sending the message, while DeviceType is the message type. Thus I can understand @adamrankin's suggestion to clarify this. The question is whether we want to change it.

Comment (adamrankin) [inline Source/igtlBindMessage.cxx]: As Christian said.

They may be consistent names, but they do not refer to consistent information. Type refers to message type, while device name refers to the device name.

This does break backwards compatibility, but I felt it was appropriate to have a properly named function. We could also do something like

    std::string GetMessageType() const;
    #ifdef OpenIGTLink_EnableV1FunctionNames
      const char* GetDeviceType();
    #endif

File: CMakeLists.txt (hunk 3 above)
Comment (adamrankin) [inline CMakeLists.txt]: I admit the dynamic function GetType() didn't make any sense to me. The type of the message is fixed to the class. The only thing I thought it might have been is a workaround for not implementing all of the GET_, STT_ and STP_ classes. If a client is sending non-standard message types, then this implementation drops the message anyways.

I would actually like to remove the m_DefaultBodyType member.

Happy to change the name, I just wanted a clean way to be able to identify the TYPE field of a message header

File: Source/igtlMessageHandler.h (hunk 5 above)
Comment (adamrankin) [inline Source/igtlMessageHandler.h]: Yes, I agree it's not backwards compatible. I am working under the assumption that v3 implementation will require people to update their code.

As above, we could implement a transition by #ifdef'ing the old function type.

With regards to memory corruption, does this not return a copy of the std::string?

File: Source/igtlutil/igtl_command.h (hunk 6 above)
Comment (adamrankin) [inline Source/igtlutil/igtl_command.h]: Why not?

2 seemed an unnecessary limit. We don't have to send 2^32 characters, but there is no harm in allowing it. Anyone who implements that kind of transfer will know the risks.

Comment (lassoan) [inline Source/igtlutil/igtl_command.h]: I see, by "- 4 bytes" you meant that uint32 is 4 bytes. I interpreted it as "minus 4 bytes" (length of the command is length-4). Probably it's better to just remove the - 4 bytes from the comment to avoid confusion.

File: CMakeLists.txt (hunk 3 above)
Comment (lassoan) [inline CMakeLists.txt]: We should do everything to minimize the changes and keep backward compatibility as much as possible. If somebody recompiles his code with a new OpenIGTLink library version then it should work as is; or if there is any non-backward compatible change then it should throw a build error. This memory corruption error will not be detected as a build error it will just cause random crashes (frequency mostly depending on CPU load, so quite difficult to find).

File: Source/igtlBindMessage.cxx (hunk 2 above)
Comment (lassoan) [inline Source/igtlBindMessage.cxx]: We should not mix essential changes with "nice-to-have" changes:
1. It'll make our changes more complex => will take more time to get reviewed and integrated
2. It'll upset the users (I still somewhat angry about how VTK's SetInput=>SetInputData method name change was forced down on throats of the application developers).

If you really want to change the name then add a new method name, add deprecation warning for the old method, and submit the whole thing as a separate pull request.

File: CMakeLists.txt (hunk 3 above)
Comment (lassoan) [inline CMakeLists.txt]: Please submit GetType-related changes as a separate pull request. All the related changes should be completed (we should remove the old mechanisms that the new mechanism is meant to replace), discussed, and then integrated. Note that probably there are 10-20 higher-priority issues than this (https://github.com/openigtlink/OpenIGTLink/issues), so I would not spend time with this refactoring now but focus on essential things.

File: Source/igtlBindMessage.cxx (hunk 2 above)
Comment (christiana) [inline Source/igtlBindMessage.cxx]: Although I agree that @adamrankin's changes would be an improvement (there are loads of stuff in this library I would like to improve), I support @lassoan's argument of minimizing the interface changes. The transition to igtlV3 should be as smooth as possible, and we will have to live with non-optimal choices simply because they are there and people are used to them.

File: CMakeLists.txt (hunk 3 above)
Comment (adamrankin) [inline CMakeLists.txt]: No problem, I am happy to revert to the existing API.

Could I add in new functionality with an #ifdef ...v3 section?

File: Source/igtlutil/igtl_command.h (hunk 6 above)
Comment (adamrankin) [inline Source/igtlutil/igtl_command.h]: Ah I see, it was in keeping with the prior lines, I will remove the size from the comments.

File: CMakeLists.txt (hunk 3 above)
Comment (adamrankin) [inline CMakeLists.txt]: Also, any recommendation for an ifdef name?

OpenIGTLink_RunTimeMessageTypeInformation?

Comment (lassoan) [inline CMakeLists.txt]: To be honest, I would save all these pseudo-RTTI related changes as they are now in a separate branch and deal with them later. First we should get all the essential features integrated into the official OpenIGTLink library with as few changes in the code as possible.

Comment (adamrankin) [inline CMakeLists.txt]: I understand that it introduces some complexity, but I think at this time it's worth it.

As it stands the definition of the IGTL specification is distributed across client and server implementations.

For example, in PLUS, it is up to us as developers to know the magic phrases that decipher message types:
https://www.assembla.com/spaces/plus/subversion/source/HEAD/trunk/PlusLib/src/PlusServer/vtkPlusOpenIGTLinkServer.cxx#ln500

MITK as well:
https://github.com/MITK/MITK/blob/master/Modules/OpenIGTLink/mitkIGTLMessageFactory.cpp#L108

And yet again in CustusX:
https://github.com/SINTEFMedtek/CustusX/blob/aef3ff22105c752276964c9db6c429adfe80c471/source/plugins/org.custusx.core.openigtlink/openigtlinkClient/protocols/cxOpenIGTLinkProtocol.cpp#L55

and in probably many others...

By adding this functionality, we are centralizing the definition of the specification to OpenIGTLink's reference implementation.

Comment (adamrankin) [inline CMakeLists.txt]: I am implementing another alternative, give me 20 minutes and we can review another approach.

Comment (lassoan) [inline CMakeLists.txt]: This is a known issue and there is already a not-yet-merged pull request with a fix (https://github.com/openigtlink/OpenIGTLink/issues/56). I just want to avoid preparing another pull request for the same issue that will not be merged because there is not enough time to properly review and test it.

Comment (adamrankin) [inline CMakeLists.txt]: Ah yes, perfect. I am slowly realizing the state of things now. I am extending the PLUS factory to use the Igtl factory.

I will perform a pass on PLUS now to see if there is any other type checking to see if there is any alternative approaches.

Comment (adamrankin) [inline CMakeLists.txt]: After a quick Skype chat with Andras I will revert any backwards compatibility breaking changes the RTTI implementation. Type checking should be delegated to the igtl::MessageFactory.

**Comment by adamrankin** [comment]
I tried to roll the changes into a single commit, but I'm not super familiar with git and the commands that I thought would work did not.

If anyone is comfortable (after merging the pull request) rolling this run of commits into a single commit, I'd love to know how you did it.
